src/renderer/atlas/shader_ps.hlsl

```diff
@@ -85,6 +87,60 @@ Output main(PSData data) : SV_Target
         weights = color.aaaa;
         break;
     }
+    case SHADING_TYPE_DASHED_LINE:
+    {
+        const bool on = frac(data.position.x / backgroundCellSize.x) < 0.5f;
+        color = on * premultiplyColor(data.color);
+        weights = color.aaaa;
+        break;
+    }
+    case SHADING_TYPE_DASHED_LINE_WIDE:
+    {
+        const bool on = frac(data.position.x / (2.0f * backgroundCellSize.x)) < 0.5f;
+        color = on * premultiplyColor(data.color);
+        weights = color.aaaa;
+        break;
+    }
+    case SHADING_TYPE_CURLY_LINE:
+    {
+        const float strokeWidthHalf = underlineWidth / 2.0f;
+        const int cellIdxY = data.position.y / backgroundCellSize.y;
+        const float cellPosY = cellIdxY * backgroundCellSize.y;
+        const float centerY = cellPosY + underlineCellOffset + strokeWidthHalf;
+        const float Pi = radians(180);
+        const float freq = 2.0f * Pi / backgroundCellSize.x;
+        const float amp = curlyLineHeight - 1.0f; // -1.0f avoids clipping at the peak
+
+        const float s = sin(data.position.x * freq);
+        const float d = abs(centerY + s * amp - data.position.y);
+        const float a = 1 - saturate(d - strokeWidthHalf);
+        color = a * premultiplyColor(data.color);
+        weights = color.aaaa;
+        break;

```

Comment:
Note to self:

This part here and in the `SHADING_TYPE_CURLY_LINE_WIDE` is quite repetitive. Can this be wrapped in a function which just gives us the values we require? 🤔

Reply:
If it's a constant value, you can use `texcoord` to smuggle integers, which seems to be all you need in this case (`1` for regular and `2` for wide underlines). Otherwise, you can of course just write a function for this. (FYI: Shaders don't really support function calls; they're all force-inlined.)

---

Try to move as many values that are constant across an entire rendering pass into the `cbuffer`. Among others, this would be `underlineCellOffset + strokeWidthHalf` and `2.0f * Pi / backgroundCellSize.x`.

---

Instead of
```hlsl
const float amp = curlyLineHeight - 1.0f; // -1.0f avoids clipping at the peak
```

I would consider increasing the size of the underline `QuadInstance` by 1px on top and bottom. I personally feel like that this would be more obvious.

Reply:
> If it's a constant value, you can use texcoord to smuggle integers, which seems to be all you need in this case (1 for regular and 2 for wide underlines).

I'm trying this but it doesn't seem to work:

```cpp
// BackendD3D.cpp

  _appendQuad() = {
        ...
        .texcoord = { static_cast<u16>(1u << horizontalShift), static_cast<u16>(1u << verticalShift) },
        ...
  };


// shader_ps.hlsl

const int horizontalScale = data.texcoord.x; 
const int verticalScale = data.texcoord.y;
```

I expect this to be due to texcoord being interpolated, but I should be able to get those integers back somehow,  just don't understand how 😅

Reply:
So, it was vertex shader which was causing the interpolation. We don't want that to happen for styled underline drawing. It needed some changes in the vertex shader and `_drawCursorForegroundSlowPath` for our texcoord plan to work, and it works 👍 It actually simplified the handling of DECDWL/DECDHL cases as well.